pkg/mutation/assign_mutator.go

```diff
@@ -123,3 +127,38 @@ func applyToToBindings(applyTos []mutationsv1alpha1.ApplyTo) []SchemaBinding {
 	}
 	return res
 }
+
+// IsValidAssign returns an error if the given assignmetadata object is not
+// semantically valid
+func IsValidAssign(assign *mutationsv1alpha1.Assign) error {
+	path, err := parser.Parse(assign.Spec.Location)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid location format")
+	}
+
+	if hasMetadataRoot(path) {
+		return fmt.Errorf("Assign can't change metadata")
+	}
+	toAssign := make(map[string]interface{})
+	err = json.Unmarshal([]byte(assign.Spec.Parameters.Assign.Raw), &toAssign)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid format for parameters.assign")
+	}
+
+	_, ok := toAssign["value"]
+	if !ok {
+		return fmt.Errorf("spec.parameters.assign must have a value field")
+	}
+	return nil
+}
+
+func hasMetadataRoot(path *parser.Path) bool {
```

Comment:
Do we want to wire in the other validation logic in #938 ?

Reply:
This is slightly different.
The other logic checks that the path is exactly metadata.labels.xxx or metadata.annotations.xxx because those are the only valid paths (so far).

Here, it's enough to have metadata as a root to make the assign object invalid.

Reply:
Good point, SGTM